tools/flash_shelly.sh

```diff
@@ -0,0 +1,324 @@
+
+  if [ $p_action != "list" ]; then
+    clear
+    echo "Host: $p_device"
+    echo "Model: $model"
+    echo "Current: $cfw"
+    echo "Latest: $lfw"
+    echo "DURL: $dlurl"
+    echo ""
+
+    cfw_V=$(convert_to_integer $cfw)
+    lfw_V=$(convert_to_integer $lfw)
+
+    if [ $official == false ] && [ $(echo "$lfw_V $cfw_V -p" | dc) -ge 1 ]; then
+      perform_flash=true
+    elif [ $official == true ] && [[ ! -z $dlurl ]]; then
+      perform_flash=true
+    else
+      perform_flash=false
+    fi
+    if [ $perform_flash == true ] && [ $dry_run == false ]; then
+      while true; do
+        read -p "Do you wish to flash $p_device to firmware version $lfw ? " yn
+        case $yn in
+          [Yy]* )  flash="yes"; break;;
+          [Nn]* ) flash="no";break;;
+          * ) echo "Please answer yes or no.";;
+        esac
+      done
+    elif [ $perform_flash == true ] && [ $dry_run == true ]; then
+      echo "Would have been converted to HomeKit firmware..."
+      read -p "Press enter to continue"
+    elif [ -z $dlurl ]; then
+      echo "$model IS NOT SUPPORTED YET..."
+      read -p "Press enter to continue"
+    else
+      echo "$p_device DOSE NOT NEED UPDATING..."
+      read -p "Press enter to continue"
+    fi
+
+    if [ "$flash" = "yes" ]; then
+      write_flash $p_device $lfw $dlurl $official
+    elif [ "$flash" = "no" ]; then
+      echo "Skipping Flash..."
+      read -p "Press enter to continue"
+    fi
+  else
+    if [ -z $dlurl ]; then
+      lfw="Not Supported"
+    fi
+    echo "Host: $p_device"
+    echo "Model: $model"
+    echo "Current: $cfw"
+    echo "Latest: $lfw"
+    echo " "
+  fi
+}
+
+function device_scan {
+  ds_device=$1
+  ds_action=$2
+  ds_do_all=$3
+  ds_dry_run=$4
+  if [ $ds_do_all == false ]; then
+    probe_info $ds_device $ds_action $ds_do_all $ds_dry_run
+  else
+    echo -e '\033[1mScanning for Shelly devices...\033[0m'
+    for device in $(timeout 2 dns-sd -B _http . | awk '/shelly/ {print $7}'); do
+      probe_info $device.local $ds_action $ds_do_all $ds_dry_run
+    done
+  fi
+}
+
+function help {
+  echo "Shelly HomeKit flashing script utility"
+  echo "Usage: $0 TODO"
+  # TODO
+}
+
+action=flash
+do_all=false
+dry_run=false
+
+while getopts ":ahln" opt; do
+  case ${opt} in
+    a )
+      do_all=true
+      ;;
+    h )
+      help
+      exit 0
+      ;;
+    l )
+      action=list
+      ;;
+    n )
+      dry_run=true
+      ;;
+    \? )
+      echo "Invalid option"
+      help
+      exit 1
+      ;;
+  esac
+done
+shift $((OPTIND -1))
+
+echo "action=$action do_all=$do_all dry_run=$dry_run"
+echo "args: $@"
+echo ""
+if [ $# == 0 -a $do_all == false ]; then
+  help
+  exit 1
+elif [[ ! -z $@ ]] && [ $do_all == true ]; then
+  help
+  exit 1
+fi
+
+# TODO: the rest of it
+release_info=$(curl -qsS -m 5 https://api.github.com/repos/mongoose-os-apps/shelly-homekit/releases/latest)
+if [[ ! -z $@ ]];then
+  for device in $@; do
+    device_scan $device $action $do_all $dry_run
```

Comment:
looks like this can be just `probe_info $device $action $dry_run`

Reply:
ye, ok